### ISSUE TITLE
Use `direnv` environment

### ISSUE THREAD

**Issue by diktomat** [zed-industries/zed#4977]
### Check for existing issues

- [X] Completed

### Describe the feature

Zed should support [direnv](https://direnv.net), to enable using a different environment (espeacially `$PATH`) for different projects. In combination with zed-industries/zed#4978 this would enable e.g. using different versions of language servers and tooling, or tooling just installed for the project instead of system-wide using Nix.

### If applicable, add mockups / screenshots to help present your vision of the feature

Inspiration from other editors:
* [wbolster/emacs-direnv](https://github.com/wbolster/emacs-direnv)
* [misuzu/direnv-subl](https://github.com/misuzu/direnv-subl)
* [direnv/direnv.vim](https://github.com/direnv/direnv.vim)
* [direnv/direnv-vscode](https://github.com/direnv/direnv-vscode)

**Comment by misuzu** [comment]
> * Direnv initializes too late, causing problems with certain extensions that rely on direnv to set the `PATH` (e.g. for utilities like language servers and linters). Ideally there was a way for the direnv extension to hook into some really early phase of initialization where it can set environmental variables before other extensions initialize.

Yep, the extension API must have support for this usecase to properly implement direnv integration. This is the reason why https://github.com/misuzu/direnv-subl is flawed - ST just doesn't have the necessary API

**Comment by szlend** [comment]
Another useful feature that I think a `direnv` extension should have is to be able to re-load the environment so you don't have to restart your editor. The way this work in vscode for example is that it changes the process environment and then restarts the extension host (and all language servers as a result).

**Comment by jeremylightsmith** [comment]
This is definitely keeping us from using zed on our project. We use direnv + nix, and we can't use tasks w/o better support.

With VSCode, it's quite easy to use direnv, you just have to launch the editor from a directory that direnv has already run with `code .`. Would it be possible to have the same behavior with zed?

**Comment by jeremylightsmith** [comment]
it seems the shared session doesn't maintain its env, because if it did, we could work around this by sourcing a script like this:

```
#!/bin/bash

if [ -z "${DIRENV_INITIALIZED}" ]; then
  echo "Loading direnv..."
  eval "$(direnv export bash)"
  export DIRENV_INITIALIZED=1
else
  echo "direnv is already loaded."
fi
```

**Comment by mrnugget** [comment]
When you open a project, we spawn a login shell in that project's root dir, get the `env` from that process, and store it to be used with some language servers.

It sounds like you have direnv/nix setup so that would work, is the problem now that it doesn't work with tasks?

**Comment by szlend** [comment]
> With VSCode, it's quite easy to use direnv, you just have to launch the editor from a directory that direnv has already run with `code .`. Would it be possible to have the same behavior with zed?

This definitely doesn't work in vscode, except for the initial window. The direnv vscode extensions fixes this and makes it somewhat usable, but it still causes cross-project env pollution from the initial vscode window.

I think how Zed does it is a step in the right direction, though it's difficult to diagnose issues when something goes wrong. It might be a good idea to document the env loading behavior and some steps to help diagnose issues (e.g. how do I inspect the window's env, what login shell Zed used to initialize the env, etc.)

**Comment by szlend** [comment]
The Nix shell also tends to set `SHELL=bash` regardless of whether you're actually in bash or not. Not sure how Zed figures out which shell to use to say if this is a problem or not though.

**Comment by mrnugget** [comment]
> I think how Zed does it is a step in the right direction, though it's difficult to diagnose issues when something goes wrong. It might be a good idea to document the env loading behavior and some steps to help diagnose issues (e.g. how do I inspect the window's env, what login shell Zed used to initialize the env, etc.)

Right now that's all printed in the logs (`zed: open logs`). So when you open Zed, you should see that we do the same in the home directory.

But yes, documentation is a good idea.

> Not sure how Zed figures out which shell to use to say if this is a problem or not though.

It uses `SHELL` and falls back to passwd lookup if `SHELL` isn't set.

It doesn't spawn `nix shell`. It uses your `$SHELL`.

This is what we do on start, in `$HOME`: https://github.com/zed-industries/zed/blob/05b658114750e9b63717ec3d05b316a927d8378e/crates/zed/src/main.rs#L309-L312

And this is what we do for the language servers, per project:

https://github.com/zed-industries/zed/blob/05b658114750e9b63717ec3d05b316a927d8378e/crates/project/src/project.rs#L11636-L11665

**Comment by musjj** [comment]
It looks like that `rust-analyzer` can now pick up the binary from the environment: https://github.com/zed-industries/zed/pull/12418. So I added this to my config:
```json
{
  "lsp": { "rust-analyzer": { "binary": { "path_lookup": true } } }
}
```
But it's still tries to download and run its own binary anyways (and failing because I'm on NixOS).

**Comment by mrnugget** [comment]
> But it's still tries to download and run its own binary anyways (and failing because I'm on NixOS).

Yes, because that commit you linked to hasn't been in any release yet.

**Comment by mersinvald** [comment]
@mrnugget it doesn't work for me without the foreground hack. If I just open the project, the env from .envrc wouldn't be loaded for anything except the integrated shell.
With foreground, both plugins and tasks work fine, as zed is loaded within the direnv-initialized initialized PTY.

> When you open a project, we spawn a login shell in that project's root dir,

Is there a way to modify what runs in this shell before editor initialization?

**Comment by mrnugget** [comment]
> @mrnugget it doesn't work for me without the foreground hack

_what_ doesn't work? Only certain language servers make use of this environment right now.

**Comment by mersinvald** [comment]
> > @mrnugget it doesn't work for me without the foreground hack
> 
> _what_ doesn't work? Only certain language servers make use of this environment right now.

I would like the editor to respect the directory env regardless of what specific plugin needs the variables it sets. In my case, the issue is with `rust-analyzer`.

**Comment by mrnugget** [comment]
> In my case, the issue is with `rust-analyzer`.

That should then be possible with the upcoming Preview release and when setting `{"rust-analyzer": {"binary": {"path_lookup": true}}}`.

**Comment by jeremylightsmith** [comment]
Alright, I want to show you where it's not working for me. You can see that when I go into my project directory, direnv runs and we are finding elixir from nix. But after launching zed, even with foreground, within a task, it finds elixir from asdf :(. Which means all my tests fail and I can't do things like `mix test`

![2024-06-11 09 52 46](https://github.com/zed-industries/zed/assets/15301/3236464c-a895-4471-8313-bb4cb1b3a036)

For VSCode, it works because the first time I run a task, it opens and fails, then I can manually type `direnv allow`, and then future runs of the task use that same environment that has been direnv'd.

**Comment by mrnugget** [comment]
> within a task

Yeah, that's not supported yet.

**Comment by zimbatm** [comment]
This is a bit related to #5347 in the sense that you want to:

1. "remote" into a per-project shell session.
2. manage the lifecycle of that session (e.g., stop/start/reload).
3. make sure every program invocation in the code has a layer of indirection to execute in that session.
4. lean on the PATH of said session to find tools instead of complicated/static lookup/installation logic.

So far, all the IDEs I have seen do this wrong as they try to resolve and execute binaries from the IDE's context. Sometimes, they allow updating the IDE's ENV, and then it breaks down when working on two different projects at the same time.

Creating this architecture is also helpful for other things, like moving to a future where project code is sandboxed correctly and doesn't have arbitrary access to all of $HOME.

**Comment by musjj** [comment]
It looks like that direnv is now officially supported: https://github.com/zed-industries/zed/blob/main/docs/src/configuring-zed.md#direnv-integration
I was trying out the latest release and was pleasantly surprised that direnv just works.

**Comment by misuzu** [comment]
> It looks like that direnv is now officially supported: https://github.com/zed-industries/zed/blob/main/docs/src/configuring-zed.md#direnv-integration I was trying out the latest release and was pleasantly surprised that direnv just works (without having to manually trigger it in the shell).

Does it work with multiple direnv environments loaded at the same time?

**Comment by jeremylightsmith** [comment]
Direnv does NOT just work. It mostly doesn't seem to work at all, and if you read the fine print on the docs, it says:

```
direnv integration currently only means that the environment variables set by a direnv configuration can be used to detect some language servers in $PATH instead of installing them.
```

I need it to find other things, like elixir in a path modified by direnv. It doesn't have the path as direnv sets it :(

**Comment by bjeanes** [comment]
> I need it to find other things, like elixir in a path modified by direnv. It doesn't have the path as direnv sets it :(

This is still true after https://github.com/zed-industries/zed/commit/0360cda5434f019ea3bda2f77c573b959249f9b9 ?

**Comment by jeremylightsmith** [comment]
I'm still on Zed 0.146.5, so looks like I need to wait till 147.2 hits? That's super exciting!

**Comment by jeremylightsmith** [comment]
Hmm, I installed zed preview, which is 147.2, and when I run this task:

```
  {
    "label": "show path",
    "command": "echo $PATH",
    "reveal": "always"
  }
```

it still shows the path without direnv's changes to it. I'm using nix, if it matters.

:(

**Comment by mrnugget** [comment]
How do you use `direnv`?

Here's what I use with Zed Preview 0.147.2:

1. Create the project:
```
mkdir ~/tmp/testing-direnv
cd ~/tmp/testing-direnv
echo export FOO=foo > .envrc
direnv allow .
```
2. Open `Zed Preview`
3. `task: Spawn`
4. Type in `echo $FOO`, hit `opt-return`
5. Task is spawned, prints `foo`

See:
![screenshot-2024-08-07-13 58 34@2x](https://github.com/user-attachments/assets/fd405786-9a6a-4e9f-8b1b-9e136fd8964f)

So, how exactly is your `direnv` setup working, how do you start Zed and how to you open that project?

**Comment by jeremylightsmith** [comment]
Thanks for taking the time to try to help me out. I think I have things working now, but I did find a couple bugs along the way.

Here's a play project that someone can hopefully repro them with:

[testing-zed.zip](https://github.com/user-attachments/files/16538453/testing-zed.zip)

It has this .envrc file:

```
export FOO=foo
export PATH=/first:$PATH:/last
```

And this .zed/tasks.json file:

```
[
  {
    "label": "show path",
    "command": "echo \"FOO=$FOO and PATH=$PATH\"",
    "reveal": "always"
  }
]
```

When I open zed and run the show path task, the behavior I see is:

1. I was using `zed .` to open zed. When I do this, I _do_ see $FOO as being defined, but I _don't_ see `/first` or `/last` on the path. Weird. When I open the application from spotlight, then I do see `/first` and `/last` on the path

2. `/last` is last on the path, but `/first` is _not_ first. not sure why, but there's a bunch of stuff that comes first, including asdf in my actual project, which breaks which elixir I'm trying to use. (I worked around it by just removing asdf for now). On the command line, `/first` is first.

I'm on an M2 Max mac running Sonoma 14.6.

**Comment by mrnugget** [comment]
Okay, let me first answer (2), since (1) is tricky and I need to dive into it.

> 2\. `/last` is last on the path, but `/first` is _not_ first. not sure why, but there's a bunch of stuff that comes first, including asdf in my actual project, which breaks which elixir I'm trying to use. (I worked around it by just removing asdf for now). On the command line, `/first` is first.

That's because when spawning tasks, we spawn an interactive shell, which then sources your `~/.zshrc` or `~/.bashrc` or what you have. 

Here, all the things that come before `/first` are things that I add in my `.zshrc`:

![screenshot-2024-08-08-10 08 41@2x](https://github.com/user-attachments/assets/20869163-c6ef-4381-b0b7-871fad22b38c)

That's reproducible when running the same command that Zed runs when executing the task:

![screenshot-2024-08-08-10 12 13@2x](https://github.com/user-attachments/assets/cbe8320f-1e1c-4cff-837e-33759794791e)

I'm not sure how to fix it yet. Workaround could be to change your shell config to detect when any of the `ZED_*` env vars are set and then not add to the PATH:

![screenshot-2024-08-08-10 16 49@2x](https://github.com/user-attachments/assets/c504b219-83f5-49de-a9a5-1b66ac883a88)

**Comment by mrnugget** [comment]
> 1. I was using `zed .` to open zed. When I do this, I _do_ see $FOO as being defined, but I _don't_ see `/first` or `/last` on the path. Weird. When I open the application from spotlight, then I do see `/first` and `/last` on the path

Okay, this is a bit of a nasty one. I think I know what's going on now:

- When you run `zed .` you are _inside_ the directory with `direnv` enabled. `direnv` leaves breadcrumbs in the environment that it's enabled: `DIRENV_DIFF`, `DIRENV_WATCHES`, `DIRENV_DIR` are all set.
- `zed` is not `Zed.app`, it's the CLI. So the CLI then spawns `Zed.app`, via macOS URL handling. See [here](https://github.com/zed-industries/zed/blob/da8d1306af711c95757fd9b7a653d7b45613c674/crates/cli/src/main.rs#L558-L567).
- `Zed.app` is then started and _inherits environment variables_ from the process that spawned it: the CLI.
  -  (You can confirm this by doing: `export ENV_VARS_SMELL_FUNKY=true && /Applications/Zed.app/Contents/MacOS/cli ~/Downloads/testing-zed`, opening a terminal inside the just-spawned Zed.app, and doing `echo $ENV_VARS_SMELL_FUNKY`)
- *But* at the same time, `Zed.app` thinks its stdout/stdin isn't connected to a TTY, as if it was spawned from the Dock. And that's correct! Because it isn't connected to the TTY that spawned it. So [this piece of logic here kicks in](https://github.com/zed-industries/zed/blob/da8d1306af711c95757fd9b7a653d7b45613c674/crates/zed/src/main.rs#L387-L398) and `Zed.app` spawns a login-shell _in your home directory_ to gather the environment as if it were spawned from the terminal (a common thing that macOS applications do that need a proper environment)
  - (You can confirm this by using `--foreground` with the CLI: `/Applications/Zed.app/Contents/MacOS/cli ~/Downloads/testing-zed --foreground`. This will leave `Zed.app` connected to the CLI and thus inheriting its stdin/stdout (and environment!) — the logic to fetch the env from the home-dir doesn't kick in and your full direnv-populated env from the shell is used.)
- This then overwrites `$PATH`! So now, we have `$PATH` from the home-dir but at the same time `DIRENV_*` environment variables set.
- So then, when you spawn the task, we try to load the environment from your project (to get the direnv environment!), but `direnv` says "nu-uh, I already ran, there's `DIRENV_*` stuff in the environment" and doesn't do anything.
  - I confirmed this by adding a bit of code locally to delete all `DIRENV_*` variables before trying to get the project environment.
- Result: you end up with the home-dir `$PATH` but the other environment variables from your CLI (like `$FOO`)

So, what do we do about it?

I think there are a bunch of options and I need to think through the trade-offs of them, but here's what we/you could do, off the top of my head:

1. We: when `Zed.app` is spawned from the CLI, we should detect that and not re-populate the environment from the home-dir. 
  - Subtask: we should explicitly forward the environment from the CLI to the process, so that we can store that on environment on the project and reuse it, even if multiple Zed windows are open. That's the best solution.
2. We could do a hack: remove `DIRENV_*` variables before spawning the login-shell in the project dir. Hacky, nasty. Temptingly easy though. Still: super nasty.
3. You: instead of doing `zed .` you could do `cd .. && zed my-dir`, so that direnv is unloaded, the env vars removed and direnv is properly loaded.
4. You: you can run `zed --foreground .` too, which should fix the issue. But I don't consider that a fix, more like a workaround.

We really should do (1) and the subtask, but we're very busy now with an upcoming launch, so I'm probably not going to get to that in the next couple of weeks.

**Comment by domenkozar** [comment]
Using 0.147.2, with direnv activated and `rust-analyzer` available in $PATH, Zed still downloads the binary on it's own and fails to run it. No error is reported in the editor.

I wish there way a way to disable auto-downloading all together and using only $PATH.

**Comment by mrnugget** [comment]
@domenkozar did you turn the PATH-lookup on for rust-analyzer? It's not the default for rust-analyzer since r-a behaves different than other language servers somehow.

```json
{
  "lsp": {
      "rust-analyzer": {
          "binary": {"path_lookup": true}
      }
  }
}
```

**Comment by domenkozar** [comment]
Same result:

```
2024-08-08T13:17:54.685708953+01:00 [INFO] starting language server "rust-analyzer", path: "/home/domen/dev/lanza.events", id: 5
2024-08-08T13:17:54.727934758+01:00 [INFO] starting language server. binary path: "/home/domen/.local/share/zed/languages/rust-analyzer/rust-analyzer-2024-08-05", working directory: "/home/domen/dev/lanza.events", args: []
2024-08-08T13:17:54.728599637+01:00 [ERROR] failed to start language server "rust-analyzer": failed to spawn command. path: "/home/domen/.local/share/zed/languages/rust-analyzer/rust-analyzer-2024-08-05", working directory: "/home/domen/dev/lanza.events", args: []
2024-08-08T13:17:54.728706493+01:00 [ERROR] server stderr: Some("")
2024-08-08T13:17:54.7288086+01:00 [ERROR] Hit 4 reinstallation attempts for "rust-analyzer"
```

**Comment by mrnugget** [comment]
It works for me:

```
$ cd tucan
direnv: loading ~/code/projects/tucan/.envrc
direnv: using flake
[...]
direnv: nix-direnv: renewed cache
direnv: export +AR +AR_FOR_TARGET +AS +AS_FOR_TARGET +CC +CC_FOR_TARGET +CONFIG_SHELL +CXX +CXX_FOR_TARGET +HOST_PATH +IN_NIX_SHELL +LD +LD_DYLD_PATH +LD_FOR_TARGET +MACOSX_DEPLOYMENT_TARGET +NIX_BINTOOLS +NIX_BINTOOLS_FOR_TARGET +NIX_BINTOOLS_WRAPPER_TARGET_HOST_aarch64_apple_darwin +NIX_BINTOOLS_WRAPPER_TARGET_TARGET_aarch64_apple_darwin +NIX_BUILD_CORES +NIX_CC +NIX_CC_FOR_TARGET +NIX_CC_WRAPPER_TARGET_HOST_aarch64_apple_darwin +NIX_CC_WRAPPER_TARGET_TARGET_aarch64_apple_darwin +NIX_CFLAGS_COMPILE +NIX_CFLAGS_COMPILE_FOR_TARGET +NIX_DONT_SET_RPATH +NIX_DONT_SET_RPATH_FOR_BUILD +NIX_ENFORCE_NO_NATIVE +NIX_HARDENING_ENABLE +NIX_IGNORE_LD_THROUGH_GCC +NIX_LDFLAGS +NIX_LDFLAGS_FOR_TARGET +NIX_NO_SELF_RPATH +NIX_STORE +NM +NM_FOR_TARGET +PATH_LOCALE +RANLIB +RANLIB_FOR_TARGET +RUST_SRC_PATH +SIZE +SIZE_FOR_TARGET +SOURCE_DATE_EPOCH +STRINGS +STRINGS_FOR_TARGET +STRIP +STRIP_FOR_TARGET +XDG_DATA_DIRS +__darwinAllowLocalNetworking +__impureHostDeps +__propagatedImpureHostDeps +__propagatedSandboxProfile +__sandboxProfile +__structuredAttrs +buildInputs +buildPhase +builder +cmakeFlags +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +dontAddDisableDepTrack +mesonFlags +name +nativeBuildInputs +out +outputs +patches +phases +preferLocalBuild +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~PATH

$ which rust-analyzer
/nix/store/x4irwblcsfq8df4z5lvq8vzwxfxfprhm-rust-default-1.77.2/bin/rust-analyzer
```

Then I open `Zed.app`, open the project, and in the logs it says:

```
2024-08-08T13:14:53.836573Z [INFO] found user-installed language server for Rust. path: "/nix/store/x4irwblcsfq8df4z5lvq8vzwxfxfprhm-rust-default-1.77.2/bin/rust-analyzer", arguments: []
2024-08-08T13:14:53.836649Z [INFO] starting language server. binary path: "/nix/store/x4irwblcsfq8df4z5lvq8vzwxfxfprhm-rust-default-1.77.2/bin/rust-analyzer", working directory: "/Users/thorstenball/code/projects/tucan", args: []
```

Can you show your `settings.json`? Or do you have a project with which to reproduce this?

**Comment by domenkozar** [comment]
```json
{
  "assistant": {
    "version": "2",
    "default_open_ai_model": null,
    "default_model": {
      "provider": "anthropic",
      "model": "claude-3-5-sonnet-20240620"
    }
  },
  "lsp": {
      "rust-analyzer": {
          "binary": {"path_lookup": true}
      }
  },
  "vim_mode": true,
  "base_keymap": "VSCode",
  "theme": "One Dark",
  "ui_font_size": 16,
  "buffer_font_size": 16
}
```

**Comment by mrnugget** [comment]
Nothing wrong there. With your settings it also works for me:

```
2024-08-08T16:07:52.718096Z [INFO] found user-installed language server for Rust. path: "/nix/store/x4irwblcsfq8df4z5lvq8vzwxfxfprhm-rust-default-1.77.2/bin/rust-analyzer", arguments: []
2024-08-08T16:07:52.718193Z [INFO] starting language server. binary path: "/nix/store/x4irwblcsfq8df4z5lvq8vzwxfxfprhm-rust-default-1.77.2/bin/rust-analyzer", working directory: "/Users/thorstenball/code/projects/tucan", args: []
```

**Comment by domenkozar** [comment]
If I launch Zed with rust-analyzer already available it works, but when it's only provisioned inside the terminal using direnv it tries to download it.

**Comment by srid** [comment]
Zed + direnv + Nix + Rust (rust-analyzer) has been consistently working well for me. Environment: macOS, without nothing (but Nix) globally installed.

The project in question: https://github.com/juspay/omnix

My Zed config:

```json
{
  "base_keymap": "VSCode",
  "load_direnv": "shell_hook",
  "lsp": { "rust-analyzer": { "binary": { "path_lookup": true } } },
  "vim_mode": true,
}

```

**Comment by srid** [comment]
Incidentally, to get [nixd](https://x.com/sridca/status/1791259775965053321) (Nix language server) working, I had to add the following to Zed config:

```json
{
  "lsp": {
    "rust-analyzer": { "binary": { "path_lookup": true } },
    "nix": { "binary": { "path_lookup": true } }
  },
}
```

<img width="688" alt="image" src="https://github.com/user-attachments/assets/68804591-9a60-4ef6-b386-a5aa0bd9a47c">

Perhaps `path_lookup` should be the default for all language servers?

**Comment by srid** [comment]
One thing I noticed is this configuration works only if you put it in your user config (`~/.config/zed/settings.json`) but not the project config (`./.zed/settings.json`) -- as [VSCode users do](https://github.com/juspay/omnix/blob/0c8d800491912625619e3e27377ba50561eea0c1/.vscode/settings.json#L2-L4). That seems like a bug?

**Comment by rawkode** [comment]
> Perhaps `path_lookup` should be the default for all language servers?

Yep! If the binary exists in the path, please just use it.

**Comment by mrnugget** [comment]
> That seems like a bug?

Not strictly a bug, but: we only allow overwriting certain settings in project-specific settings. But yeah, agree, this should probably be one.

**Comment by mrnugget** [comment]
> Yep! If the binary exists in the path, please just use it.

I agree, but last time I did that, lots of `rust-analyzer` stuff broke, I need to look into this again.

**Comment by srid** [comment]
@rome-user `0.149.3` on macOS ARM 15.0 beta.

<img width="249" alt="image" src="https://github.com/user-attachments/assets/fa246c04-2174-440e-a192-7d24cc02529c">

I definitely have to enable `path_lookup` for each LSP to get it to work (can confirm for rust and nixd).

**Comment by domenkozar** [comment]
A good step forward would be to have a global `path_lookup` flag that each LSP implementation needs to respect.

**Comment by DrewBurkhart** [comment]
In case it helps anyone else, this was my solution/workaround.

---

***tl;dr***

What was tripping me up is that it was saying `Hit 4 reinstallation attempts for "rust-analyzer"` but _I think_ it was really 4 failed attempts to ***start*** rust-analyzer. Once I installed it via `rustup`, it worked just fine. That step wasn't required in VS Code so I just hadn't expected to need it. Not sure if that's expected or not.

---

<details>

<summary>
My Problem when using nix + direnv + rust-analyzer (essentially just a restatement of earlier messages here)
</summary>

In various projects that work as is in VS code with no RA-related settings, I'm having trouble getting RA to run. I've added
```
{
  "load_direnv": "shell_hook",
  "lsp": {
    "rust-analyzer": {
      "binary": { "path_lookup": true }
    }
  }
}
```
to `~/.config/zed/settings.json` but it appears as though one of two things happens depending on how I start Zed (which matches what was explained above):

1. If I open via CLI (`zed .`) it doesn't load the env in the terminal and rust-analyzer seems to start ok but then errors with "failed to load workspace" in the logs (with no other apparent context).

2. If I open via spotlight it loads the env correctly but the logs indicate that it's failing to install rust-analyzer repeatedly (same as @domenkozar noted above)
```
[INFO] found user-installed language server for Rust. path: "/nix/store/dlipzjnxcypn8sz8kgaqn7hw26j4shki-rustup-1.26.0/bin/rust-analyzer", arguments: []
[INFO] starting language server. binary path: "/nix/store/dlipzjnxcypn8sz8kgaqn7hw26j4shki-rustup-1.26.0/bin/rust-analyzer", working directory: "/Users/me/Development/proj", args: []
[ERROR] cannot read LSP message headers
[ERROR] failed to start language server "rust-analyzer": oneshot canceled
[ERROR] server stderr: Some("error: 'rust-analyzer' is not installed for the toolchain '1.80.0-aarch64-apple-darwin'\n")
[ERROR] Hit 4 reinstallation attempts for "rust-analyzer"
```
This is the same behavior as when I follow the steps @mrnugget mentioned above
> You: instead of doing zed . you could do cd .. && zed my-dir, so that direnv is unloaded, the env vars removed and direnv is properly loaded.
You: you can run zed --foreground . too, which should fix the issue. But I don't consider that a fix, more like a workaround.

</details>

---

<details>

<summary>
My Solution/Workaround
</summary>

Basically I eventually just followed the error I'd overlooked in the logs which states
```
[ERROR] server stderr: Some("error: 'rust-analyzer' is not installed for the toolchain '1.80.0-aarch64-apple-darwin'\n")
```
So I ran
```
rustup component add rust-analyzer --toolchain 1.80.0-aarch64-apple-darwin
```
and then ran
```
cd .. && zed proj
```

Works great now!

</details>

**Comment by deng232** [comment]
in nixos. when I launch zed inside an manually launched development shell. ```direnv``` work properly. but with terminal direnv shell hook setup, zed termianl shows getting inside development shell but it is not develop environment. both "shell_hook" and "direct" setting behave the same.

**Comment by mrnugget** [comment]
@DrewBurkhart thanks so much for that debugging! That's going to help.

**Comment by domenkozar** [comment]
Currently, some LSPs call `workspace.which(binary)` to get the binary if it exists, while others use `path_lookup` settings (which is off by default).

Ideally, the extension API would provide `find_executable(name)` function and then have a global setting for path_lookup configurability for uniform experience.

Once that lands I'll write documentation for [devenv](https://devenv.sh/editor-support/zed/).

**Comment by mrnugget** [comment]
> Currently, some LSPs call `workspace.which(binary)` to get the binary if it exists, while others use `path_lookup` settings (which is off by default).

I'm in the process of turning `path_lookup` for rust-analyzer on by default. https://github.com/zed-industries/zed/pull/17926

**Comment by xuoe** [comment]
I no longer experience this issue with the latest version (`0.164.2`). Previously, Zed wouldn't pick up project-specific LSP binaries; it would prefer either global/user ones or, if those were missing, those tied to extensions (installed in `$XDG_DATA_HOME`). Tried most things people here recommended to no avail.

Now it loads the expected binary (here `templ`) even if Zed is started from a `.desktop` entry:

```
[INFO] starting language server process. binary path: "/nix/store/5xdjrsqz69k324826bkn4514bhxqgzf9-templ/bin/templ", working directory: "/home/_/x/y/z", args: ["lsp"]
```

Skimming [the release notes](https://zed.dev/releases/stable/0.164.2), I suspect it may have something to do with #21039.

**Comment by mrnugget** [comment]
> Unfortunately, it seems to work correctly only when reloading a project. If Zed auto-loads a project on startup, it fails to detect the binaries, but if I toggle between projects, it detects the binaries both in the secondary project and the initial one.

Is it possible that you start Zed from the CLI there? In that case, that environment takes precedence over everything else we load.

**Comment by xuoe** [comment]
> Is it possible that you start Zed from the CLI there? In that case, that environment takes precedence over everything else we load.

Yes. The result of doing that is the same as opening a project from an existing instance.

Here are the scenarios, possibly non-exhaustive, where the LSP binaries are un/detected:

| Undetected                                                                                  | Detected                                                                      |
| -------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------- |
| `$ zed my/project`                                                                           | `$ cd my/project && zed .`                                                    |
| `$ zed` and waiting for Zed to auto-load the last project                                    | `$ zed --new` and opening a project                                           |
| similarly, launching Zed from an application launcher, where Zed auto-loads the last project | `$ zed` and opening a project, regardless of one being auto-loaded on startup |

After comparing Zed logs and shell environments across all these scenarios, I've concluded that, in the *Undetected* cases, the main Zed process simply doesn't load the direnv environment. However, in both *Undetected* and *Detected* cases, the Zed terminal shell is spawned with the correct environment. Toggling between `shell_hook` and `direct` had no apparent effect.

Relevant log lines with the actual paths elided:

```
[INFO] set environment variables from shell:/run/current-system/sw/bin/fish, path: ...
[INFO] using project environment variables from CLI. PATH=...
```

The paths in question, present in all the *Detected* cases above, yet missing in all the *Undetected* ones, are as follows:

- `<project>/.direnv/bin`
- project-specific `/nix/store/*` paths generated after `flake.nix`  is applied

---

Looking at [the code](https://github.com/zed-industries/zed/blob/664468d468ff52d74fb6e99849543f5bc01c1267/crates/project/src/environment.rs), I'm not sure why a login shell must be invoked just to grab the environment. To my mind, it is the user's responsibility to set their session variables as they see fit.

In other words, the environment available to Zed at invocation time, regardless of how it's invoked, that should be the initial environment. The environment generated by `direnv export json` in the project directory is the environment diff that should be applied to the initial one and passed to terminal shells as well<sup>[1]</sup>. This would streamline the process and make it easier to reason about.

<sup>[1]</sup> For the sake of completeness (and correctness), this would likely have to be bidirectional (i.e., terminal shell -> Zed should work as well), in the event that direnv picks up changes to `.envrc` or `flake.nix`; this latter step would possibly require LSP server path checks and restarts, if paths differ. Tasks would likely also be affected.

**Comment by szlend** [comment]
> Looking at [the code](https://github.com/zed-industries/zed/blob/664468d468ff52d74fb6e99849543f5bc01c1267/crates/project/src/environment.rs), I'm not sure why a login shell must be invoked just to grab the environment. To my mind, it is the user's responsibility to set their session variables as they see fit.

My understanding is this is necessary to get a fresh base environment, otherwise the environment from the first window pollutes the environment of subsequent windows you open, since they all share the same base environment (they're not fresh processes but fork of the original process).

**Comment by uncenter** [comment]
Is there a reason this issue is still open? Seems like support is working fine now for the most part, other issues should probably be bug reports since the core functionality should be all there? I got confused thinking Zed didn't support direnv at all because this feature request was open.

**Comment by mrnugget** [comment]
Yeah, agree. Let's close this and open separate issues with MREs in them.

**Comment by olebedev** [comment]
Hi @mrnugget, I still have this issue with Zed 0.166.1. Where can I post full report? Thank you!

**Comment by FlafyDev** [comment]
Some environments are more than just the environment variables. Like FHS nix environment also modify the root directories like `/bin`. Is there a solution for those? Maybe allow to run multiple instances of Zed? Is that possible in release mode?
